cpp/src/arrow/compute/kernels/scalar_temporal.cc

```diff
@@ -528,13 +576,18 @@ struct Strftime {
   static Result<Strftime> Make(KernelContext* ctx, const DataType& type) {
     const StrftimeOptions& options = StrftimeState::Get(ctx);
 
-    auto timezone = GetInputTimezone(type);
     // This check is due to surprising %c behavior.
     // See https://github.com/HowardHinnant/date/issues/704
     if ((options.format.find("%c") != std::string::npos) && (options.locale != "C")) {
       return Status::Invalid("%c flag is not supported in non-C locales.");
     }
-    if (timezone.empty()) {
+    std::string timezone;
+    bool is_timestamp = type.id() == Type::TIMESTAMP;
+    if (is_timestamp) {
```

Comment:
You could make this a compile-time check:
https://github.com/apache/arrow/blob/78ac1322de9d31fdf89b8db7d1bd786928fe3a32/cpp/src/arrow/type_traits.h#L703

Then presumably any timestamp-related code would get optimized out for the non-timestamp specializations.

Reply:
Good point! Changed, please check if it makes sense.